File: code/ch02/2.8.1.fractionToDecimal.cpp
#include <cstdlib>
#include <cmath>
#include <string>
#include <map>
using namespace std;

typedef long long ll;

class Solution {
public:
  string fractionToDecimal(ll numerator,
      ll denominator) {
    string res("");
    if (double(numerator) / double(denominator) < 0.0) res += "-";
    numerator = llabs(numerator);
    denominator = llabs(denominator);
    ll n = numerator / denominator;
    ll remainder = numerator % denominator;
    res += to_string(n);
    if (!remainder) return res;
    res += ".";
    int idx = res.size();
    map<ll, int> seen;
    while (remainder && !seen.count(remainder)) {
      seen[remainder] = idx++;
      n = (remainder * 10LL) / denominator;
      remainder = (remainder * 10LL) % denominator;
      res += to_string(n);
    }
    if (remainder) {
      res.insert(seen[remainder], 1, '(');
      res += ")";
    }
    return res;
  }
};

File: code/ch02/2.8.1.fractionToDecimal.java
import java.util.*;
import java.lang.Math;

public class Solution {
  public String fractionToDecimal(long numerator,
      long denominator) {
    String res = "";
    if ((double)numerator / (double)denominator < 0.0) res = res + "-";
    numerator = Math.abs(numerator);
    denominator = Math.abs(denominator);
    long n = numerator / denominator;
    long remainder = numerator % denominator;
    res = res + String.valueOf(n);
    if (remainder == 0) return res;
    res = res + ".";
    int idx = res.length();
    Map<Long, Integer> seen = new HashMap<Long,
      Integer>();
    while (remainder != 0 && seen.get(remainder) == null) {
      seen.put(remainder, idx);
      idx = idx + 1;
      n = (remainder * 10) / denominator;
      remainder = (remainder * 10) % denominator;
      res = res + String.valueOf(n);
    }
    if (remainder != 0) {
      StringBuffer ano_str =  new StringBuffer(res);
      ano_str.insert(seen.get(remainder), "(");
      res = ano_str.toString();
      res += ")";
    }
    return res;
  }
}
